ipfs-pinning-service.yaml

```diff
@@ -246,10 +262,13 @@ components:
       type: object
       required:
         - cid
+        - providers
       properties:
         cid:
-          description: CID to be pinned
+          description: CID to be pinned recursively
           type: string
+        providers:
+          $ref: '#/components/schemas/Providers'
```

Comment:
If we're providing a user back with the multiaddresses their nodes need to connect to, do we need to have them pass in their own nodes as well? This property might not be needed here.

If we're making the provider node response mandatory and the applications implementing this are planning on utilizing that info to connect to the pinning service's node as part of their workflow, users needing to provide their own multiaddress could probably be left out of the spec.

While Pinata currentlys allow users to pass in their own multiaddresses, after thinking this through more, relying on the users to connect their own nodes to ours seems like the better solution because it cuts down on user error. If a user's node is behind a NAT, or if they accidentally pick the wrong multiaddress (a private ip for example) we might not be able to connect to it which creates friction. In the case of us providing back our node's multiaddress, theoretically it should work 100% of the time.

Reply:
> If we're providing a user back with the multiaddresses their nodes need to connect to, do we need to have them pass in their own nodes as well? This property might not be needed here.

Indeed, `Pin.providers` is not as crucial for data transfer as `PinStatus.providers`, but its still a hint for pinning service to act on when present, and a metadata that user may want to persist along with Pin (eg. from which PeerID the data originated). 

`Pin.providers` can be left empty if user is pinning data that is publicly available and wants DHT-based discovery to occur. 

But.. I wonder if we should make `PinService.providers` length min 1 to ensure pinning services always return at least one multiaddr?

Reply:
If we keep Pin.providers it should have a much lower max than 1000. I assume the Pinning Services have some rate limiting built in, but saying "hey, dial these 1k addresses" is not good. Maybe treat this like a DHT shortcut and cap it at 20(K).

> I wonder if we should make PinService.providers length min 1 to ensure pinning services always return at least one multiaddr

I think this would be useful.

Reply:
Separated them in 6832ad4a5bb2143ab11b5397041553978bce9ca9 and added limit of 20 multiaddrs.

Reply:
20 might even be overkill here. Our limit is 5 currently and I don't think we've ever actually had 5 passed in to our API.

Reply:
The max of 20 is just a failsafe against user/configuration error.
It is perfectly fine to pass/return only one or two.